### PR TITLE
[WIP] [Netlify] status badge

### PR DESCRIPTION
**Pull request by AnandChowdhary**
For #3129

### CODE REVIEW

File: services/netlify/netlify.tester.js [hunk 1]
```diff
@@ -0,0 +1,15 @@
+'use strict'
+
+const { isColorMatcher } = require('../color-matcher')
+const t = (module.exports = require('../tester').createServiceTester())
+
+t.create('netlify (valid)')
+  .get('/0028c6a7-b7ae-49f6-b847-917b40b5b13f')
```
Comment (calebcartwright) [inline services/netlify/netlify.tester.js]: Make sure to include the `.json` extension on these tests

Comment (AnandChowdhary) [reply services/netlify/netlify.tester.js]: Has been added, still looks like something's wrong.

**Comment by AnandChowdhary** [comment]
@paulmelnikow @PyvesB I'm not sure I entirely understand the 404 test which is failing (I didn't do a good job copying it from Travis). Could someone maybe have a quick look? 😄

**Comment by PyvesB** [comment]
I currently don't have access to a computer, I'll take a look sometime tomorrow is no one has done so by then. :wink

**Comment by calebcartwright** [comment]
@AnandChowdhary - The default error message on a 404 error is `not found`. When you want to customize that to a different message (such as `app not found`), you've got to include the `errorMessages` arg with your desired message. I added an inline comment on your fetch function where you'll need to include that

File: services/netlify/netlify.service.js [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+        staticPreview: renderColorStatusBadge({ status: '#0f0' }),
+      },
+    ]
+  }
+
+  async fetch({ appId }) {
+    const url = `https://api.netlify.com/api/v1/badges/${appId}/deploy-status`
+    return this._requestSvg({
+      schema,
+      url,
```
Comment (calebcartwright) [inline services/netlify/netlify.service.js]:
```suggestion
      errorMessages: {
        404: 'app not found',
      },
```

**Comment by AnandChowdhary** [comment]
Hey @paulmelnikow, any chance you could have a look at the Heroku logs and see what's going wrong? 😄

**Comment by calebcartwright** [comment]
Heroku logs from review app:
```
2019-03-09T13:10:56.220471+00:00 app[web.1]: at Module.require (internal/modules/cjs/loader.js:705:19)
2019-03-09T13:10:56.220472+00:00 app[web.1]: at require (internal/modules/cjs/helpers.js:14:16)
2019-03-09T13:10:56.220474+00:00 app[web.1]: at Object.<anonymous> (/app/services/color-matcher.js:9:22)
2019-03-09T13:10:56.220476+00:00 app[web.1]: at Module._compile (internal/modules/cjs/loader.js:799:30)
2019-03-09T13:10:56.220477+00:00 app[web.1]: at Object.Module._extensions..js (internal/modules/cjs/loader.js:810:10)
2019-03-09T13:10:56.220479+00:00 app[web.1]: at Module.load (internal/modules/cjs/loader.js:666:32)
2019-03-09T13:10:56.220481+00:00 app[web.1]: at tryModuleLoad (internal/modules/cjs/loader.js:606:12)
2019-03-09T13:10:56.220482+00:00 app[web.1]: at Function.Module._load (internal/modules/cjs/loader.js:598:3)
2019-03-09T13:10:56.220484+00:00 app[web.1]: at Module.require (internal/modules/cjs/loader.js:705:19)
2019-03-09T13:10:56.220486+00:00 app[web.1]: at require (internal/modules/cjs/helpers.js:14:16)
2019-03-09T13:10:56.220487+00:00 app[web.1]: at Object.<anonymous> (/app/services/netlify/netlify.service.js:5:36)
2019-03-09T13:10:56.220489+00:00 app[web.1]: at Module._compile (internal/modules/cjs/loader.js:799:30)
2019-03-09T13:10:56.220490+00:00 app[web.1]: at Object.Module._extensions..js (internal/modules/cjs/loader.js:810:10)
2019-03-09T13:10:56.220492+00:00 app[web.1]: at Module.load (internal/modules/cjs/loader.js:666:32)
2019-03-09T13:10:56.220497+00:00 app[web.1]: at tryModuleLoad (internal/modules/cjs/loader.js:606:12)
2019-03-09T13:10:56.220499+00:00 app[web.1]: at Function.Module._load (internal/modules/cjs/loader.js:598:3)
2019-03-09T13:10:56.220501+00:00 app[web.1]: at Module.require (internal/modules/cjs/loader.js:705:19)
2019-03-09T13:10:56.220502+00:00 app[web.1]: at require (internal/modules/cjs/helpers.js:14:16)
2019-03-09T13:10:56.220504+00:00 app[web.1]: at servicePaths.forEach.servicePath (/app/core/base-service/loader.js:26:20)
2019-03-09T13:10:56.220506+00:00 app[web.1]: at Array.forEach (<anonymous>)
2019-03-09T13:10:56.220522+00:00 app[web.1]: at loadServiceClasses (/app/core/base-service/loader.js:25:16)
2019-03-09T13:10:56.220524+00:00 app[web.1]: at Server.registerServices (/app/core/server/server.js:189:5)
2019-03-09T13:10:56.220526+00:00 app[web.1]: at Server.start (/app/core/server/server.js:271:10)
2019-03-09T13:10:56.220528+00:00 app[web.1]: at /app/server.js:29:18
2019-03-09T13:10:56.220530+00:00 app[web.1]: at Object.<anonymous> (/app/server.js:34:3)
2019-03-09T13:10:56.220532+00:00 app[web.1]: at Module._compile (internal/modules/cjs/loader.js:799:30)
2019-03-09T13:10:56.220534+00:00 app[web.1]: at Object.Module._extensions..js (internal/modules/cjs/loader.js:810:10)
2019-03-09T13:10:56.220535+00:00 app[web.1]: at Module.load (internal/modules/cjs/loader.js:666:32)
2019-03-09T13:10:56.220537+00:00 app[web.1]: at tryModuleLoad (internal/modules/cjs/loader.js:606:12)
2019-03-09T13:10:56.220538+00:00 app[web.1]: at Function.Module._load (internal/modules/cjs/loader.js:598:3)
2019-03-09T13:10:56.220540+00:00 app[web.1]: at Function.Module.runMain (internal/modules/cjs/loader.js:862:12)
2019-03-09T13:10:56.220542+00:00 app[web.1]: at internal/main/run_main_module.js:21:11 code: 'MODULE_NOT_FOUND' }
2019-03-09T13:10:56.344796+00:00 heroku[web.1]: State changed from starting to crashed
2019-03-09T13:10:56.324600+00:00 heroku[web.1]: Process exited with status 1
2019-03-09T14:05:38.788811+00:00 heroku[router]: at=error code=H10 desc="App crashed" method=GET path="/" host=shields-staging-pr-3178.herokuapp.com request_id=45c7b413-a312-454e-8406-d7bd506fa6cf fwd="85.144.7.181" dyno= connect= service= status=503 bytes= protocol=https
2019-03-09T14:05:39.651663+00:00 heroku[router]: at=error code=H10 desc="App crashed" method=GET path="/favicon.ico" host=shields-staging-pr-3178.herokuapp.com request_id=81fd1ba0-7bfe-414e-aab0-95660cf2620b fwd="85.144.7.181" dyno= connect= service= status=503 bytes= protocol=https
2019-03-09T14:05:58.544092+00:00 heroku[router]: at=error code=H10 desc="App crashed" method=GET path="/" host=shields-staging-pr-3178.herokuapp.com request_id=6e46a1db-ab96-4f80-bc0c-57cfada03dc9 fwd="85.144.7.181" dyno= connect= service= status=503 bytes= protocol=https
2019-03-09T14:05:59.733405+00:00 heroku[router]: at=error code=H10 desc="App crashed" method=GET path="/favicon.ico" host=shields-staging-pr-3178.herokuapp.com request_id=661715cc-da74-4a75-971c-ef2722f2dec8 fwd="85.144.7.181" dyno= connect= service= status=503 bytes= protocol=https
2019-03-09T14:39:40.575356+00:00 heroku[web.1]: State changed from crashed to starting
2019-03-09T14:39:45.560243+00:00 heroku[web.1]: Starting process with command `node server`
2019-03-09T14:39:48.389335+00:00 app[web.1]: raven@2.6.4 alert: no DSN provided, error reporting disabled
2019-03-09T14:39:48.392160+00:00 app[web.1]: Configuration:
2019-03-09T14:39:48.395348+00:00 app[web.1]: { bind: { address: '0.0.0.0', port: '51749' },
2019-03-09T14:39:48.395351+00:00 app[web.1]: metrics: { prometheus: { enabled: false } },
2019-03-09T14:39:48.395353+00:00 app[web.1]: ssl: { isSecure: false },
2019-03-09T14:39:48.395354+00:00 app[web.1]: cors: { allowedOrigin: [] },
2019-03-09T14:39:48.395356+00:00 app[web.1]: persistence: { dir: './private' },
2019-03-09T14:39:48.395358+00:00 app[web.1]: services:
2019-03-09T14:39:48.395359+00:00 app[web.1]: { github:
2019-03-09T14:39:48.395361+00:00 app[web.1]: { baseUri: 'https://api.github.com/',
2019-03-09T14:39:48.395363+00:00 app[web.1]: debug: { enabled: false, intervalSeconds: 200 } },
2019-03-09T14:39:48.395364+00:00 app[web.1]: trace: false },
2019-03-09T14:39:48.395366+00:00 app[web.1]: profiling: { makeBadge: false },
2019-03-09T14:39:48.395367+00:00 app[web.1]: cacheHeaders: { defaultCacheLengthSeconds: 120 },
2019-03-09T14:39:48.395369+00:00 app[web.1]: rateLimit: true,
2019-03-09T14:39:48.395370+00:00 app[web.1]: handleInternalErrors: true,
2019-03-09T14:39:48.395372+00:00 app[web.1]: fetchLimit: '10MB' }
2019-03-09T14:39:49.401656+00:00 app[web.1]: 0309143949 FsTokenPersistence configured with /app/private/github-user-tokens.json
2019-03-09T14:39:49.403144+00:00 app[web.1]: 0309143949 Server is starting up: http://0.0.0.0:51749/
2019-03-09T14:39:49.873898+00:00 app[web.1]: { Error: Cannot find module 'nearest-color'
2019-03-09T14:39:49.873917+00:00 app[web.1]: at Function.Module._resolveFilename (internal/modules/cjs/loader.js:649:15)2019-03-09T14:39:49.873919+00:00 app[web.1]: at Function.Module._load (internal/modules/cjs/loader.js:575:25)
2019-03-09T14:39:49.873922+00:00 app[web.1]: at Module.require (internal/modules/cjs/loader.js:705:19)
2019-03-09T14:39:49.873924+00:00 app[web.1]: at require (internal/modules/cjs/helpers.js:14:16)
2019-03-09T14:39:49.873925+00:00 app[web.1]: at Object.<anonymous> (/app/services/color-matcher.js:9:22)
2019-03-09T14:39:49.873927+00:00 app[web.1]: at Module._compile (internal/modules/cjs/loader.js:799:30)
2019-03-09T14:39:49.873930+00:00 app[web.1]: at Object.Module._extensions..js (internal/modules/cjs/loader.js:810:10)
2019-03-09T14:39:49.873931+00:00 app[web.1]: at Module.load (internal/modules/cjs/loader.js:666:32)
2019-03-09T14:39:49.873933+00:00 app[web.1]: at tryModuleLoad (internal/modules/cjs/loader.js:606:12)
2019-03-09T14:39:49.873934+00:00 app[web.1]: at Function.Module._load (internal/modules/cjs/loader.js:598:3)
2019-03-09T14:39:49.873936+00:00 app[web.1]: at Module.require (internal/modules/cjs/loader.js:705:19)
2019-03-09T14:39:49.873938+00:00 app[web.1]: at require (internal/modules/cjs/helpers.js:14:16)
2019-03-09T14:39:49.873940+00:00 app[web.1]: at Object.<anonymous> (/app/services/netlify/netlify.service.js:5:36)
2019-03-09T14:39:49.873941+00:00 app[web.1]: at Module._compile (internal/modules/cjs/loader.js:799:30)
2019-03-09T14:39:49.873943+00:00 app[web.1]: at Object.Module._extensions..js (internal/modules/cjs/loader.js:810:10)
2019-03-09T14:39:49.873944+00:00 app[web.1]: at Module.load (internal/modules/cjs/loader.js:666:32)
2019-03-09T14:39:49.873946+00:00 app[web.1]: at tryModuleLoad (internal/modules/cjs/loader.js:606:12)
2019-03-09T14:39:49.873947+00:00 app[web.1]: at Function.Module._load (internal/modules/cjs/loader.js:598:3)
2019-03-09T14:39:49.873949+00:00 app[web.1]: at Module.require (internal/modules/cjs/loader.js:705:19)
2019-03-09T14:39:49.873951+00:00 app[web.1]: at require (internal/modules/cjs/helpers.js:14:16)
2019-03-09T14:39:49.873953+00:00 app[web.1]: at servicePaths.forEach.servicePath (/app/core/base-service/loader.js:26:20)
2019-03-09T14:39:49.873955+00:00 app[web.1]: at Array.forEach (<anonymous>)
2019-03-09T14:39:49.873956+00:00 app[web.1]: at loadServiceClasses (/app/core/base-service/loader.js:25:16)
2019-03-09T14:39:49.873958+00:00 app[web.1]: at Server.registerServices (/app/core/server/server.js:189:5)
2019-03-09T14:39:49.873960+00:00 app[web.1]: at Server.start (/app/core/server/server.js:271:10)
2019-03-09T14:39:49.873962+00:00 app[web.1]: at /app/server.js:29:18
2019-03-09T14:39:49.873963+00:00 app[web.1]: at Object.<anonymous> (/app/server.js:34:3)
2019-03-09T14:39:49.873965+00:00 app[web.1]: at Module._compile (internal/modules/cjs/loader.js:799:30)
2019-03-09T14:39:49.873966+00:00 app[web.1]: at Object.Module._extensions..js (internal/modules/cjs/loader.js:810:10)
2019-03-09T14:39:49.873968+00:00 app[web.1]: at Module.load (internal/modules/cjs/loader.js:666:32)
2019-03-09T14:39:49.873970+00:00 app[web.1]: at tryModuleLoad (internal/modules/cjs/loader.js:606:12)
2019-03-09T14:39:49.873972+00:00 app[web.1]: at Function.Module._load (internal/modules/cjs/loader.js:598:3)
2019-03-09T14:39:49.873974+00:00 app[web.1]: at Function.Module.runMain (internal/modules/cjs/loader.js:862:12)
2019-03-09T14:39:49.873975+00:00 app[web.1]: at internal/main/run_main_module.js:21:11 code: 'MODULE_NOT_FOUND' }
2019-03-09T14:39:49.989978+00:00 heroku[web.1]: Process exited with status 1
2019-03-09T14:39:50.016807+00:00 heroku[web.1]: State changed from starting to crashed
2019-03-09T16:48:49.551341+00:00 heroku[router]: at=error code=H10 desc="App crashed" method=GET path="/" host=shields-staging-pr-3178.herokuapp.com request_id=0f503720-5af4-438a-9044-0009f44d7223 fwd="76.187.101.175" dyno= connect= service= status=503 bytes= protocol=https
2019-03-09T16:48:49.907098+00:00 heroku[router]: at=error code=H10 desc="App crashed" method=GET path="/favicon.ico" host=shields-staging-pr-3178.herokuapp.com request_id=9e148e80-c831-41e7-b7c1-228d3dbe8273 fwd="76.187.101.175" dyno= connect= service= status=503 bytes= protocol=https
2019-03-09T16:50:54.014707+00:00 heroku[router]: at=error code=H10 desc="App crashed" method=GET path="/" host=shields-staging-pr-3178.herokuapp.com request_id=6fbd7c39-470e-4cd8-bd0c-c0cc40038cdc fwd="76.187.101.175" dyno= connect= service= status=503 bytes= protocol=https
2019-03-09T16:50:54.298978+00:00 heroku[router]: at=error code=H10 desc="App crashed" method=GET path="/favicon.ico" host=shields-staging-pr-3178.herokuapp.com request_id=ce7bd6d6-5570-4801-aa7b-6312105cf063 fwd="76.187.101.175" dyno= connect= service= status=503 bytes= protocol=https
```

**Comment by calebcartwright** [comment]
I've not looked at this PR in detail yet given that it's still marked as WIP, but I believe you'll need `nearest-color` to be a prod dependency, and not a dev dependency

**Comment by paulmelnikow** [comment]
For what it's worth, it's awesome that we're now getting those errors when deploy apps are created!

Can I ask, why are we adding this dependency? Can the color be matched more simply? If the service tests, or better yet the schema, are written in such a way that they fail if the colors happen to change in the future, we can easily update the code when that happens. If that happens more than once, that'd be a good time to solve the problem using a dependency.

**Comment by AnandChowdhary** [comment]
@paulmelnikow I've pushed a new version with strict color check (no color matcher), but I'm having trouble with the tests for the example. 😞

**Comment by paulmelnikow** [comment]
This pull request **introduces 1 alert** when merging 6c2a46ef7e2219c074fe326ea2db06ec362819ce into cbcf980182b02382bae4deec177f07be1909abef - [view on LGTM.com](https://lgtm.com/projects/g/badges/shields/rev/pr-01c5ffbe6d0f9f149a3f4f1673f9984b712af43e)

**new alerts:**

* 1 for Wrong use of &#39;this&#39; for static method

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by paulmelnikow** [comment]
@AnandChowdhary I fixed a couple things, and where this is at now, the service tests are still failing. It looks like the data that's coming out of the SVG parser is a color, but it's being matched using a schema that wants a build status string.

Probably you want the schema to read something like this:

```js
const schema = Joi.object({
  message: Joi.equal('007A70', 'A88100', … ).required(),
}).required()
```

If you run `npm run test:services:trace` you can see some debug output related to the problem.

The other failing test looks to be transient / unrelated.

**Comment by paulmelnikow** [comment]
Hi @AnandChowdhary, it would be great to get this merged! Are you up for digging in or would you prefer someone else wrap this up?

**Comment by AnandChowdhary** [comment]
I'm on it! I'll push another update tonight to this branch and we can have a look at the final tests and the merge-ability of the branch. Thanks for being patient! 😄

**Comment by paulmelnikow** [comment]
I'd love to get this in. Feel free to reopen it when you're ready!

Anyone else reading this who want this implemented, you can grab the commits from the pull request branch using `git fetch origin pull/3178/head:pr-3178`, address the comments, and reopen.

**Comment by RDIL** [comment]
Any word on this?

**Pull request closed without merging**